### PR TITLE
Added an Editor with upload/saving functionalities

### PR DESCRIPTION
**Pull request by davidli3100**
This is kind of two tasks in one, so I would appreciate it if I could get credit for both. Or I could just extend upon this and make a dedicated view for uploading a file. 

Used `react-ace` to create the editor, and a few libraries to implement saving/reading of files

### CODE REVIEW

**Comment by KeenWarrior** [comment]
@davidli3100 sorry for the inconvenience but we needed to change based folder structure. can you please rebase your commits?

**Comment by anshumanv** [comment]
Thanks for the massive amount of work on this, please rebase as per the current folder structure. Thanks!

**Review by divyanshu-rawat — COMMENTED** [review]
Looks good, but do resolve merge conflicts, and do rebase.

**Comment by davidli3100** [comment]
Alright, rebasing tonight

File: src/Components/customEditor/customEditor.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        this.aceRef = this.refs.aceEditor;
+    }
+
+    onFileUpload = (file) => {
+        var fileName = file.name;
+        var fileType = file.type;
+        this.setState({fileName: fileName, fileType: fileType})
+        var reader = new FileReader();
+
+        console.log(file)
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: Please do not leave console statements in the code.

File: src/Components/customEditor/customEditor.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+        reader.onload = (e) => {
+            this.aceRef = this.refs.aceEditor;
+            console.log(this.aceRef)
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: would be great if you remove this.

File: src/Components/customEditor/customEditor.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            fileType: 'text/javascript'
+        }
+    }
+
+    componentDidMount() {
+        this.aceRef = this.refs.aceEditor;
+    }
+
+    onFileUpload = (file) => {
+        var fileName = file.name;
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: if filename is not getting changes, which I believe is not, then make it a `const`

File: src/Components/customEditor/customEditor.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        var fileType = file.type;
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: same here.

File: src/Components/customEditor/customEditor.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            this.aceRef = this.refs.aceEditor;
+            console.log(this.aceRef)
+            this.aceRef.editor.setValue(e.target.result);
+        }
+
+        reader.readAsText(file, "UTF-8");
+    }
+
+    onFileDownload = () => {
+        console.log(this.state)
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: would be great, if you remove console statement.

File: src/Components/customEditor/customEditor.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        var file = new File([this.aceRef.editor.getValue()], this.state.fileName, {type: this.state.fileType+";charset=utf-8"});
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: if it is not changing make it `const`.

File: src/Components/customEditor/customEditor.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        FileSaver.saveAs(file);
+    }
+
+    // openModal = () => {
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: don"t leave comments.

File: src/Components/toolbar/toolbar.css [hunk 8]
```diff
@@ -0,0 +1,3 @@
+.fileUploadBox {
+    margin: 10px 20px
```
Comment (divyanshu-rawat) [inline src/Components/toolbar/toolbar.css]: I think you missed semi-colon here.

File: src/Views/editor.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import Toolbar from '../Components/toolbar/toolbar'
+
+export default class Editor extends Component {
+    
+
+    
+    render() {
+        return (
+            <Fragment>
+                {/* <Toolbar/> */}
```
Comment (divyanshu-rawat) [inline src/Views/editor.js]: what is this here?

**Review by divyanshu-rawat — CHANGES_REQUESTED** [review]

File: src/Components/customEditor/customEditor.js [hunk 10]
```diff
@@ -27,38 +27,24 @@ export default class CustomEditor extends Component {
     }
 
     onFileUpload = (file) => {
-        var fileName = file.name;
-        var fileType = file.type;
+        const fileName = file.name;
+        const fileType = file.type;
         this.setState({fileName: fileName, fileType: fileType})
         var reader = new FileReader();
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: reader is not getting changed right? can you check if we can also make it a `const`

File: src/Components/customEditor/customEditor.js [hunk 11]
```diff
@@ -27,38 +27,24 @@ export default class CustomEditor extends Component {
...
-            console.log(this.aceRef)
             this.aceRef.editor.setValue(e.target.result);
         }
 
         reader.readAsText(file, "UTF-8");
     }
 
     onFileDownload = () => {
-        console.log(this.state)
         var file = new File([this.aceRef.editor.getValue()], this.state.fileName, {type: this.state.fileType+";charset=utf-8"});
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: It is still not changed to `const`

File: src/Views/editor.js [hunk 12]
```diff
@@ -8,10 +8,7 @@ export default class Editor extends Component {
     
     render() {
         return (
-            <Fragment>
-                {/* <Toolbar/> */}
-                <CustomEditor/>
-            </Fragment>
+            <CustomEditor/>
```
Comment (divyanshu-rawat) [inline src/Views/editor.js:10]: Looks good!

**Review by divyanshu-rawat — CHANGES_REQUESTED** [review]

File: src/Components/customEditor/editor.css [hunk 13]
```diff
@@ -0,0 +1,3 @@
+#mainEditor {
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/editor.css]: what is the purpose of this css `id`, remove it if you are not making use of it.

File: src/Components/toolbar/toolbar.js [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        fileUploaded: false,
+        files: [],
+        fileCount: 0
+    }
+
+    handleUpload = () => {
+        this.setState({isModalOpen: true})
+    }
+
+    _closeModal = () => {
```
Comment (divyanshu-rawat) [inline src/Components/toolbar/toolbar.js]: any specific reason for naming the functions starting with `_`, prefer normal camelCase approach as you did in other cases.

File: src/Components/toolbar/toolbar.js [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+        this.setState({isModalOpen: false})
+    }
+
+    _items = [
```
Comment (divyanshu-rawat) [inline src/Components/toolbar/toolbar.js]: same here any reason for starting the name with `_`

File: src/Components/toolbar/toolbar.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+                    ref={ref => this.pond = ref}
+                    files={this.state.files}
+                    className="fileUploadBox"
+                    onupdatefiles={fileItems => {
+                        this.setState({
+                          files: fileItems.map(fileItem => fileItem.file)
+                        });
+                        if(fileItems.length > 0) {
+                            this.props.onFileUpload(fileItems[0].file)
+                            if(this.state.fileCount === 0) {
```
Comment (divyanshu-rawat) [inline src/Components/toolbar/toolbar.js]: for readability point of view, you can also declare fileCount above return, like this.

`const { fileCount } = this.state; `

 or 

` const { state: { fileCount }} = this;`

File: src/Components/toolbar/toolbar.js [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+          // This needs an ariaL    abel since it's icon-only
+          ariaLabel: 'Info',
+          iconOnly: true,
+          iconProps: { iconName: 'Info' },
+          onClick: () => console.log('Info')
+        }
+    ];
+    
+    render() {
+        return (
```
Comment (divyanshu-rawat) [inline src/Components/toolbar/toolbar.js:98]: I think above `return` you can declare all **state variables** to make it more readable.
Something like this.

`const { fileCount, isModalOpen, files } = this.state; `

**or** 

`const { state: { fileCount, isModalOpen, files }} = this;`

Comment (divyanshu-rawat) [reply src/Components/toolbar/toolbar.js:98]: It is totally upto you which way you prefer. :)

File: src/Components/toolbar/toolbar.js [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+            <div>
```
Comment (divyanshu-rawat) [inline src/Components/toolbar/toolbar.js:99]: you can also use `<Fragment />` if you want, it's totally upto you.

File: src/Components/customEditor/customEditor.js [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    componentDidMount() {
+        this.aceRef = this.refs.aceEditor;
+    }
+
+    onFileUpload = (file) => {
+        const fileName = file.name;
+        const fileType = file.type;
+        this.setState({fileName: fileName, fileType: fileType})
```
Comment (divyanshu-rawat) [inline src/Components/customEditor/customEditor.js]: I think you can also re-write this as.

`this.setState({ fileName, fileType });`

**Review by divyanshu-rawat — COMMENTED** [review]
I have added all general suggestion that will make code more red-able.
Let me know your thoughts on my feedback :).

**Comment by divyanshu-rawat** [comment]
@davidli3100 are you still working on it?

**Comment by davidli3100** [comment]
Yeah I am. A few more readability issues left

**Comment by davidli3100** [comment]
Updated the suggested readability changes. Thanks for the feedback!

**Review by divyanshu-rawat — APPROVED** [review]
LGTM 👍

**Comment by davidli3100** [comment]
Cool! Let me know if there are any more React/JS based tasks left over,

**Comment by divyanshu-rawat** [comment]
https://deploy-preview-1--react-online-editor.netlify.com/ looks good to me.

**Pull request merged**